File: Recon/python/ecalPreselectionSkimmer.py
"""Configuration for PreselectionProcessor

Sets all the default parameters that high so it leads to no preselection.

Attributes:
------------- 

use_rechits: bool
    If True, use rechit-based preselection. If False, use veto-based preselection (default: False)

Rechit-based mode parameters:
ecal_rec_hit_coll: string
    Collection Name for ecal rechits
ecal_rec_hit_pass: string
    Pass Name for ecal rechits

Veto-based mode parameters:
ecal_veto_name: string
    Collection Name for veto object
ecal_veto_pass: string
    Pass Name for veto object
ecal_mip_name: string
    Collection Name for mip result object
ecal_mip_pass: string
    Pass Name for mip result object

Shared parameters (used in both modes):
summed_det_max: double
    Max value for summed det (veto mode) or total rechit energy sum (rechit mode, MeV)
 n_readout_hits_max: int
    Max value for num readout hits (veto mode) or num rechits (rechit mode)
summed_tight_iso_max: double
   Max value for summed tigh iso (veto mode only)
ecal_back_energy_max: double
    Max value for ecal back energy (veto mode only)
 shower_rms_max: double 
    Max value for shower rms
 shower_y_std_max: double
   Max value for shower rms in Y
shower_x_std_max: double
    Max value for shower rms in X
max_cell_dep_max: double
    Max value for maximal cell deposition
std_layer_hit_max: int
    Max value for std layer hits
n_straight_tracks_max: int
    Max value for num straight tracks
bdt_disc_min: double
    Min value for the BDT disc variable
fiducial_level: int
    0: don't care if it's fiducial or not, 
    1: keep fiducial events only, 
    2: keep non-fid events only



Examples
--------
    from LDMX.Recon.ecalPreselectionSkimmer import EcalPreselectionSkimmer
    
    # Veto-based mode (default):
    ecal_pres_skimmer = EcalPreselectionSkimmer()
    p.sequence.append( ecal_pres_skimmer )
    
    # Rechit-based mode:
    ecal_pres_skimmer = EcalPreselectionSkimmer()
    ecal_pres_skimmer.use_rechits = True
    ecal_pres_skimmer.ecal_rec_hit_coll = "EcalRecHits"
    ecal_pres_skimmer.ecal_rec_hit_pass = ""
    ecal_pres_skimmer.summed_det_max = 4000.0  # MeV (total rechit energy)
    ecal_pres_skimmer.n_readout_hits_max = 90  # num rechits
    p.sequence.append( ecal_pres_skimmer )
"""

from LDMX.Framework import ldmxcfg

class EcalPreselectionSkimmer(ldmxcfg.Producer) :
    """Configuration for an ECAL-based pre-selection skimmer"""

    def __init__(self, name = "ecalPreselectionSkimmer") :
        super().__init__(name,'recon::EcalPreselectionSkimmer','Recon')

        # Mode selection (default: rechit-based)
        self.use_rechits = True
        
        # Rechit-based parameters
        self.ecal_rec_hit_coll = "EcalRecHits"
        self.ecal_rec_hit_pass = ""
        
        # Veto-based parameters
        self.ecal_veto_name = "EcalVeto"
        self.ecal_veto_pass = ""
        self.ecal_mip_name = 'EcalMipInfo'
        self.ecal_mip_pass = ''

        # Cut values
        self.summed_det_max = 4000.0  # MeV
        self.summed_tight_iso_max = 9999.
        self.ecal_back_energy_max = 9999.
        self.n_readout_hits_max = 90
        self.shower_rms_max = 9999.
        self.shower_y_std_max = 9999.
        self.shower_x_std_max = 9999.
        self.max_cell_dep_max = 9999.
        self.std_layer_hit_max = 9999
        self.n_straight_tracks_max = 9999
        self.bdt_disc_min = 0.
        self.fiducial_level = 0

File: Recon/src/Recon/Skims/EcalPreselectionSkimmer.cxx
/**
 * @file EcalPreselectionSkimmer.cxx
 * @brief Processor used to pre-select events for the ECAL studies
 * @author Tamas Almos Vami (UCSB)
 */

#include "Recon/Skims/EcalPreselectionSkimmer.h"

namespace recon {

EcalPreselectionSkimmer::EcalPreselectionSkimmer(const std::string &name,
                                                 framework::Process &process)
    : framework::Producer(name, process) {}

void EcalPreselectionSkimmer::configure(framework::config::Parameters &ps) {
  ecal_veto_name_ = ps.get<std::string>("ecal_veto_name");
  ecal_veto_pass_ = ps.get<std::string>("ecal_veto_pass");
  ecal_mip_name_ = ps.get<std::string>("ecal_mip_name");
  ecal_mip_pass_ = ps.get<std::string>("ecal_mip_pass");
  summed_det_max_ = ps.get<double>("summed_det_max");              // MeV
  summed_tight_iso_max_ = ps.get<double>("summed_tight_iso_max");  // MeV
  ecal_back_energy_max_ = ps.get<double>("ecal_back_energy_max");  // MeV
  n_readout_hits_max_ = ps.get<int>("n_readout_hits_max");
  shower_rms_max_ = ps.get<double>("shower_rms_max");
  shower_y_std_max_ = ps.get<double>("shower_y_std_max");
  shower_x_std_max_ = ps.get<double>("shower_x_std_max");
  max_cell_dep_max_ = ps.get<double>("max_cell_dep_max");  // MeV
  std_layer_hit_max_ = ps.get<int>("std_layer_hit_max");
  n_straight_tracks_max_ = ps.get<int>("n_straight_tracks_max");
  bdt_disc_min_ = ps.get<double>("bdt_disc_min");
  fiducial_level_ = ps.get<int>("fiducial_level");

  return;
}

void EcalPreselectionSkimmer::produce(framework::Event &event) {
  bool passed_preselection{false};
  bool fiducial_decision{true};
  const auto &ecal_veto{
      event.getObject<ldmx::EcalVetoResult>(ecal_veto_name_, ecal_veto_pass_)};
  const auto &mip_result{
      event.getObject<ldmx::EcalMipResult>(ecal_mip_name_, ecal_mip_pass_)};
  // Boolean to if we skim for fiducial / nonfiducial
  fiducial_decision = (fiducial_level_ == 0 ||
                       (fiducial_level_ == 1 && ecal_veto.getFiducial()) ||
                       (fiducial_level_ == 2 && !ecal_veto.getFiducial()));

  // Boolean to check if we pass preselection
  passed_preselection =
      (ecal_veto.getSummedDet() < summed_det_max_) &&
      (ecal_veto.getSummedTightIso() < summed_tight_iso_max_) &&
      (ecal_veto.getEcalBackEnergy() < ecal_back_energy_max_) &&
      (ecal_veto.getNReadoutHits() < n_readout_hits_max_) &&
      (ecal_veto.getShowerRMS() < shower_rms_max_) &&
      (ecal_veto.getYStd() < shower_y_std_max_) &&
      (ecal_veto.getXStd() < shower_x_std_max_) &&
      (ecal_veto.getMaxCellDep() < max_cell_dep_max_) &&
      (ecal_veto.getStdLayerHit() < std_layer_hit_max_) &&
      (mip_result.getNStraightTracks() < n_straight_tracks_max_) &&
      (ecal_veto.getDisc() > bdt_disc_min_) && fiducial_decision;

  // Tell the skimmer to keep or drop the event based on whether preselection
  // passed
  if (passed_preselection) {
    ldmx_log(debug) << "This event passed preselection!";
    setStorageHint(framework::HINT_SHOULD_KEEP);
  } else {
    setStorageHint(framework::HINT_SHOULD_DROP);
  }
  // Add the boolean to the event
  event.add("EcalPreselectionDecision", passed_preselection);
}
}  // namespace recon

DECLARE_PRODUCER(recon::EcalPreselectionSkimmer);
